src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/MetadataService.kt

```diff
@@ -0,0 +1,189 @@
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.retry
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.suspendUntil
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.elasticapi.getManagedIndexMetadata
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.transport.action.updateindexmetadata.UpdateManagedIndexMetaDataAction
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.transport.action.updateindexmetadata.UpdateManagedIndexMetaDataRequest
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.util.managedIndexMetadataIndexRequest
+import com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement.util.revertISMMetadataID
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.OpenForTesting
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.ExceptionsHelper
+import org.elasticsearch.action.DocWriteRequest
+import org.elasticsearch.action.bulk.BackoffPolicy
+import org.elasticsearch.action.bulk.BulkItemResponse
+import org.elasticsearch.action.bulk.BulkRequest
+import org.elasticsearch.action.bulk.BulkResponse
+import org.elasticsearch.action.support.master.AcknowledgedResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.block.ClusterBlockException
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.common.unit.TimeValue
+import org.elasticsearch.index.Index
+import org.elasticsearch.rest.RestStatus
+import java.lang.Exception
+
+/**
+ * When all nodes have same version IM plugin, MetadataService starts to move
+ * metadata from cluster state to metadata index
+ */
+@OpenForTesting
+class MetadataService(
+    private val client: Client,
+    private val clusterService: ClusterService,
+    private val skipExecution: SkipExecution,
+    private val imIndices: IndexManagementIndices
+) {
+    private val logger = LogManager.getLogger(javaClass)
+
+    @Volatile private var runningLock = false // in case 2 moveMetadata() process running
+
+    private val successIndexedIndices = mutableSetOf<metadataDocID>()
+    private var failedToIndexIndices = mutableMapOf<metadataDocID, BulkItemResponse.Failure>()
+    private var failedToCleanIndices = mutableListOf<Index>()
+
+    private var counter = 0
+    @Volatile private var finishFlag = false
+    // used in coordinator sweep to cancel scheduled process
+    fun isFinished() = finishFlag
+
+    @Volatile
+    private var retryPolicy =
+        BackoffPolicy.constantBackoff(TimeValue.timeValueMillis(50), 3)
+
+    // just implement the business logic here, no need to do those listen to settings
+    // schedule a runnable
+    suspend fun moveMetadata() {
+        if (runningLock) {
+            logger.info("There is a move metadata process running...")
+            return
+        }
+        runningLock = true
```

Comment:
Are you sure no branching path below that ends up with you taking a runningLock and never undoing it by accident? To protect against that you could use a try finally block instead

Reply:
makes sense, wrap `runningLock = false` in a finally block